Modules/_sqlite/connection.c

```diff
@@ -897,21 +897,121 @@ print_or_clear_traceback(callback_context *ctx)
     }
 }
 
+// Return a borrowed reference to built-in str().
+static PyObject *
+fetch_str_builtin(callback_context *ctx)
+{
+    pysqlite_state *state = ctx->state;
+    if (state->bltin_str != NULL) {
+        return state->bltin_str;
+    }
+
+    /* Both PyEval_GetBuiltins and PyDict_GetItemString return
+     * borrowed references. */
+    PyObject *bltins = PyEval_GetBuiltins();
+    if (bltins == NULL) {
+        return NULL;
+    }
+    PyObject *str = PyDict_GetItemString(bltins , "str");
+    if (str == NULL) {
+        return NULL;
+    }
+
+    /* Store a strong reference in the module state;
+     * return a borrowed reference to the caller. */
+    state->bltin_str = Py_NewRef(str);
+    return str;
+}
+
+static PyObject *
+get_exception_message(callback_context *ctx, PyObject *exc)
+{
+    PyObject *str = fetch_str_builtin(ctx);  // Borrowed ref.
+    if (str == NULL) {
+        return NULL;
+    }
+    PyObject *res = PyObject_CallFunction(str, "(O)", exc);
+    if (res == NULL) {
+        return NULL;
+    }
+    return res;
+}
+
+static char *
+build_error_message(const char *preamble, const char *message)
```

Comment:
It may be simpler to use `PyUnicode_Format()`.

Also, it is worth to include also the type name of the original error, not only its message.

Reply:
Thanks, those are good suggestions.

Reply:
Since we're going to end up with a `char *` and not a `PyObject *`, I think it is best to use the C stdlib here.

Reply:
Try to use `PyUnicode_Format()`. The code will be much simpler and easier to extend. You do not even need to use `PyObject_Str()` explicitly.

Reply:
I'll do that, but I'm not happy with how the error messages themselves turn out; I need to tweak this more before it is ready.

Reply:
I finally found time to look at this again; using `PyUnicode_FromFormat` was indeed a good suggestion. Thanks!